tool/src/c2/ty.rs

```diff
@@ -222,57 +233,59 @@ impl<'ccx, 'tcx: 'ccx, 'header> TyGenContext<'ccx, 'tcx, 'header> {
                 ]
             }
             Type::Slice(hir::Slice::Primitive(b, p)) if !is_struct => {
-                let constness = self.cx.formatter.fmt_constness(b.mutability);
                 let prim = self.cx.formatter.fmt_primitive_as_c(*p);
+                let ptr_type = self.cx.formatter.fmt_ptr(&prim, b.mutability);
                 vec![
                     (
-                        format!("{constness}{prim}*").into(),
+                        format!("{ptr_type}").into(),
                         format!("{param_name}_data").into(),
                     ),
                     ("size_t".into(), format!("{param_name}_len").into()),
                 ]
             }
             _ => {
-                let ty = self.gen_ty_name(ty);
+                let ty = self.gen_type_name(ty, is_struct);
                 vec![(ty, param_name)]
             }
         }
     }
 
     // Generate the C code for referencing a particular type.
     // Handles adding imports and such as necessary
-    fn gen_ty_name<P: TyPosition>(&mut self, ty: &Type<P>) -> Cow<'ccx, str> {
+    fn gen_type_name<P: TyPosition>(&mut self, ty: &Type<P>, is_decl: bool) -> Cow<'ccx, str> {
```

Comment:
issue: i really don't want that is_decl param

Reply:
#302